atomic_reactor/util.py

```diff
@@ -1672,3 +1672,177 @@ def chain_get(d, path, default=None):
         except (IndexError, KeyError):
             return default
     return obj
+
+
+class DockerfileImages(object):
+    def __init__(self, dfp_parents):
+        self._original_parents = deepcopy(dfp_parents) if dfp_parents else []
+        self._custom_parent_image = False
+        self._custom_base_image = False
+        self._base_from_scratch = False
+        self._source_registry = None
+        self._organization = None
+        self._original_base_image = self._original_parents[-1] if self._original_parents else None
+        self._pullable_parents = []
+        self._local_parents = []
+        self._create_pullable()
+
+    @property
+    def original_parents(self):
+        """
+        provides unmodified parent_images from Dockerfile
+        """
+        return self._original_parents
+
+    @property
+    def original_base_image(self):
+        """
+        provides unmodified base image from Dockerfile
+        """
+        return self._original_base_image
+
+    @property
+    def base_image(self):
+        """
+        provides local tag for base_image, or base image with registry and enclosed
+        when we don't know local tag yet
+        """
+        if self.base_from_scratch:
+            return SCRATCH_FROM
+
+        if self._local_parents and self._local_parents[0]:
+            return deepcopy(self._local_parents[0])
+        elif self._pullable_parents and self._pullable_parents[0]:
+            return deepcopy(self._pullable_parents[0])
+        else:
+            raise KeyError('base_image')
+
+    @property
+    def base_image_key(self):
+        """
+        provides base image with registry and enclosed
+        (original base image with added registry and enclosed)
+        """
+        if self.base_from_scratch:
+            return SCRATCH_FROM
+
+        if self._pullable_parents and self._pullable_parents[0]:
+            return deepcopy(self._pullable_parents[0])
+        else:
+            raise KeyError('base_image')
+
+    @property
+    def base_from_scratch(self):
+        return self._base_from_scratch
+
+    @property
+    def custom_base_image(self):
+        return self._custom_base_image
+
+    @property
+    def custom_parent_image(self):
+        return self._custom_parent_image
```

Comment:
Can we have lazy evaluation here?

Reply:
value will be set once class is instantiated and it isn't suppose to change